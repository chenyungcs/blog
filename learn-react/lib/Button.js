/**
 * Button
 * @author    : yunchen
 * @createdAt : 06/03/2018
 */
import React from 'react'
import PropTypes from 'prop-types'

const Button = ({ onClick, className, children }) => (
  <button
    onClick={onClick}
    className={className}
    type="button"
  >
    {children}
  </button>
)

Button.propTypes = {
  onClick: PropTypes.func,
  className: PropTypes.string,
  size: PropTypes.oneOf(['smaller', 'normal', 'large']), // 枚举
  children: PropTypes.node
}

Button.defaultProps = {
  className: 'button-link button-active'
}

export default Button
