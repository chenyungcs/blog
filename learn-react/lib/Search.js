/**
 * Search
 * @author    : yunchen
 * @createdAt : 06/03/2018
 */
import React, { Component } from 'react'

export default class Search extends Component {
  componentDidMount () {
    if (this.input) {
      this.input.focus()
    }
  }

  render () {
    const {
      value,
      onChange,
      onSubmit,
      children = '搜索'
    } = this.props

    return (
      <form onSubmit={onSubmit}>
        <input
          type="text"
          value={value}
          onChange={onChange}
          ref={(node) => (this.input = node)}
        />
        <button type="submit">
          {children}
        </button>
      </form>
    )
  }
}

