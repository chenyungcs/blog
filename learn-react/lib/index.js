/**
 * index
 * @author    : yunchen
 * @createdAt : 06/03/2018
 */
export { default as Button } from './Button'
export { default as Table } from './Table'
export { default as Search } from './Search'
