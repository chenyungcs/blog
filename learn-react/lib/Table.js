/**
 * Table
 * @author    : yunchen
 * @createdAt : 06/03/2018
 */
import React, { Component } from 'react'
import Button from './Button'
import PropTypes from 'prop-types'

const isSearched = searchTerm => item => item.objectID

export const type = {name: 'Table'}

export default class Table extends Component {
  static propTypes = {
    list: PropTypes.arrayOf(PropTypes.object).isRequired, // 组合类型
    pattern: PropTypes.string,
    onDismiss: PropTypes.func
  }

  static defaultProps = {
    list: []
  }

  render () {
    const { list, pattern, onDismiss } = this.props

    return (
      <section>
        <ul className="table">
          {list.filter(isSearched(pattern)).map((item, index) => (
            <li key={item.objectID} className="table-row">
              <span style={{width: '40%'}}>
                <a href={item.url}>{index + 1} - {item.title || '无'}</a>
              </span>
              <span style={{width: '30%'}}>{item.author}</span>
              <span style={{width: '10%'}}>{item.num_comments}</span>
              <span style={{width: '10%'}}>{item.points}</span>
              <span style={{width: '10%'}}>
                <Button
                  className="button-inline"
                  onClick={() => onDismiss(item.objectID)}
                >
                  Dismiss
                </Button>
              </span>
            </li>
          ))}
        </ul>
      </section>
    )
  }
}
