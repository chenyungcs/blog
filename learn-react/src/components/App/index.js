/**
 * index
 * @author    : yunchen
 * @createdAt : 06/03/2018
 */
import React, { Component } from 'react'
import { Button, Table, Search } from '../../../lib'
import fetch from 'isomorphic-fetch'

const Loading = () => (
  <div style={{padding: '0 10px'}}>Loading...</div>
)

export default class App extends Component {
  constructor (props) {
    super(props)

    this.state = {
      searchTerm: '',
      page: 0,
      list: [],
      error: '',
      loading: false
    }
    this.lock = false

    this.onSearchChange = this.onSearchChange.bind(this)
    this.fetchData = this.fetchData.bind(this)
    this.onSearchSubmit = this.onSearchSubmit.bind(this)
    this.onDismiss = this.onDismiss.bind(this)
  }

  fetchData (keywords = '', page = 0) {
    if (this.state.loading) { return }
    this.setState({loading: true})

    fetch(`https://hn.algolia.com/api/v1/search?query=${keywords}&page=${page}`)
      .then(resp => resp.json())
      .then(({ hits, page }) => this.setData({page, list: hits}))
      .then(() => this.setState({loading: false}))
      .catch(error => {
        this.setState({
          error: error.message,
          loading: false
        })
      })
  }

  setData ({ page = 0, list = [] } = {}) {
    this.setState({
      page,
      list: [
        ...this.state.list,
        ...list
      ]
    })
  }

  componentDidMount () {
    this.fetchData()
  }

  onSearchChange (e) {
    this.setState({searchTerm: e.target.value})
  }

  onSearchSubmit (event) {
    const { searchTerm } = this.state
    this.setState({page: 0, list: []})
    this.fetchData(searchTerm)
    event.preventDefault()
  }

  onDismiss (id) {
    const isNotId = (item) => item.objectID !== id
    const updateData = this.state.list.filter(isNotId)
    this.setState({
      list: updateData
    })
  }

  render () {
    const { searchTerm, list, page = 0, error, loading } = this.state

    return (
      <section className="page">
        { error
          ? <p>{error}</p>
          : <Search
            value={searchTerm}
            onChange={this.onSearchChange}
            onSubmit={this.onSearchSubmit}
          />
        }
        {list && <Table list={list} pattern={searchTerm} onDismiss={this.onDismiss} />}
        { loading
          ? <Loading />
          : <Button onClick={() => this.fetchData(searchTerm, page + 1)}>下一页 {page + 1}</Button>
        }
      </section>
    )
  }
}

