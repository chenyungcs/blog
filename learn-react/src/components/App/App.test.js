/**
 * App.test
 * @author    : yunchen
 * @createdAt : 06/03/2018
 */
import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import App from '../App'
import { Button } from '../../../lib'

Enzyme.configure({adapter: new Adapter()})

describe('App', () => {
  it('render without crashing', () => {
    const div = document.createElement('div')
    ReactDOM.render(<App />, div)
  })

  // 快照测试
  test('has a void snapshot', () => {
    const component = renderer.create(<App />)
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe('Button', () => {
  const props = {
    className: 'btn'
  }

  const element = shallow(<Button {...props}>搜索</Button>)

  it('text is "搜索"', () => {
    expect(element.text()).toBe('搜索')
  })

  it('className is "btn"', () => {
    expect(element.is('.btn')).toBe(true)
  })

  it('type is button', () => {
    expect(element.is('button[type="button"]')).toBe(true)
  })
})
