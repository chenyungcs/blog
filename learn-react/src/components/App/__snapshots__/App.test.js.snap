// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App has a void snapshot 1`] = `
<section
  className="page"
>
  <form
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
    <button
      type="submit"
    >
      搜索
    </button>
  </form>
  <section>
    <ul
      className="table"
    />
  </section>
  <button
    className="button-link button-active"
    onClick={[Function]}
    type="button"
  >
    下一页 
    1
  </button>
</section>
`;
